### ISSUE TITLE
Installation issue on Freebsd 

### ISSUE THREAD

**Issue by fritz-hh** [jbarlow83/OCRmyPDF#375]
**Describe the issue**
Installation of ocrmypdf on freebsd (jail) fails using pip3 (python3,7) 
```
Failed to build pikepdf
Could not build wheels for pikepdf which use PEP 517 and cannot be installed directly
```
(see full log below)

**To Reproduce**
Run `pip3 install ocrmypdf`

**Log**

```
root@ocrmypdf:/ # pip3 install ocrmypdf
Collecting ocrmypdf
  Using cached https://files.pythonhosted.org/packages/ec/0c/2e5aa3083748a21ff18d6ceb778d93a794a07ba6f90b72bf278adfb68ce6/ocrmypdf-8.2.3-py35-none-any.whl
Collecting chardet<4,>=3.0.4 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting pikepdf<2,>=1.1.0 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/d0/de/87be0b2b9dd5626d87df75f7f5ee04276cde661095eb4f854193eeaa32c4/pikepdf-1.1.0.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting pdfminer.six==20181108; sys_platform != "darwin" (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/8a/fd/6e8746e6965d1a7ea8e97253e3d79e625da5547e8f376f88de5d024bacb9/pdfminer.six-20181108-py2.py3-none-any.whl
Collecting cffi>=1.9.1 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/64/7c/27367b38e6cc3e1f49f193deb761fe75cda9f95da37b67b422e62281fcac/cffi-1.12.2.tar.gz
Collecting ruffus>=2.7.0 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/18/24/05e8fe590d08bd9e6122c6a87425ff741c79edf01d2873f92028e860e547/ruffus-2.8.1.tar.gz
Collecting img2pdf<0.4,>=0.3.0 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/e0/c6/7cd14232a1b10bf884c12daf3626afb76c4f60b52ae0eb23ce1519542ae4/img2pdf-0.3.3.tar.gz
Collecting reportlab>=3.3.0 (from ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/f1/db/980bf931c2c9f2931d3ed539430b6465495ab4365d76303bc008226a50f3/reportlab-3.5.18.tar.gz
Collecting defusedxml>=0.5.0 (from pikepdf<2,>=1.1.0->ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/87/1c/17f3e3935a913dfe2a5ca85fa5ccbef366bfd82eb318b1f75dadbf0affca/defusedxml-0.5.0-py2.py3-none-any.whl
Collecting lxml>=4.0 (from pikepdf<2,>=1.1.0->ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/7d/29/174d70f303016c58bd790c6c86e6e86a9d18239fac314d55a9b7be501943/lxml-4.3.3.tar.gz
Collecting sortedcontainers (from pdfminer.six==20181108; sys_platform != "darwin"->ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/13/f3/cf85f7c3a2dbd1a515d51e1f1676d971abe41bba6f4ab5443240d9a78e5b/sortedcontainers-2.1.0-py2.py3-none-any.whl
Collecting pycryptodome (from pdfminer.six==20181108; sys_platform != "darwin"->ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/c0/1e/5398467ff0fcb39dbb23412c5ea9c9b2db2b659282e49d4073daae952247/pycryptodome-3.8.1.tar.gz
Collecting six (from pdfminer.six==20181108; sys_platform != "darwin"->ocrmypdf)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting pycparser (from cffi>=1.9.1->ocrmypdf)
  Downloading https://files.pythonhosted.org/packages/68/9e/49196946aee219aead1290e00d1e7fdeab8567783e83e1b9ab5585e6206a/pycparser-2.19.tar.gz (158kB)
    100% |████████████████████████████████| 163kB 1.9MB/s
Collecting Pillow (from img2pdf<0.4,>=0.3.0->ocrmypdf)
  Downloading https://files.pythonhosted.org/packages/81/1a/6b2971adc1bca55b9a53ed1efa372acff7e8b9913982a396f3fa046efaf8/Pillow-6.0.0.tar.gz (29.5MB)
    100% |████████████████████████████████| 29.5MB 246kB/s
Building wheels for collected packages: pikepdf
  Building wheel for pikepdf (PEP 517) ... error
  Complete output from command /usr/local/bin/python3.7 /usr/local/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmp8uin0bl9:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7
  creating build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  copying src/pikepdf/_version.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  copying src/pikepdf/_methods.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  copying src/pikepdf/__init__.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  copying src/pikepdf/_cpphelpers.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  copying src/pikepdf/objects.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf
  creating build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf/models
  copying src/pikepdf/models/matrix.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf/models
  copying src/pikepdf/models/image.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf/models
  copying src/pikepdf/models/__init__.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf/models
  copying src/pikepdf/models/metadata.py -> build/lib.freebsd-11.2-RELEASE-p3-amd64-3.7/pikepdf/models
  running build_ext
  creating tmp
  cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -O2 -pipe -fstack-protector -fno-strict-aliasing -fPIC -I/usr/local/include/python3.7m -c /tmp/tmpq0s7bm7i.cpp -o tmp/tmpq0s7bm7i.o -std=c++14
  cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -O2 -pipe -fstack-protector -fno-strict-aliasing -fPIC -I/usr/local/include/python3.7m -c /tmp/tmpx2raeo76.cpp -o tmp/tmpx2raeo76.o -fvisibility=hidden
  building 'pikepdf._qpdf' extension
  creating build/temp.freebsd-11.2-RELEASE-p3-amd64-3.7
  creating build/temp.freebsd-11.2-RELEASE-p3-amd64-3.7/src
  creating build/temp.freebsd-11.2-RELEASE-p3-amd64-3.7/src/qpdf
  cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -O2 -pipe -fstack-protector -fno-strict-aliasing -fPIC -I/usr/local/include/python3.7m -I/root/.local/include/python3.7m -I/usr/local/include/python3.7m -c src/qpdf/shims.cpp -o build/temp.freebsd-11.2-RELEASE-p3-amd64-3.7/src/qpdf/shims.o -DVERSION_INFO="1.1.0" -std=c++14 -fvisibility=hidden
  cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -O2 -pipe -fstack-protector -fno-strict-aliasing -fPIC -I/usr/local/include/python3.7m -I/root/.local/include/python3.7m -I/usr/local/include/python3.7m -c src/qpdf/object.cpp -o build/temp.freebsd-11.2-RELEASE-p3-amd64-3.7/src/qpdf/object.o -DVERSION_INFO="1.1.0" -std=c++14 -fvisibility=hidden
  src/qpdf/object.cpp:14:10: fatal error: 'qpdf/Constants.h' file not found
  #include <qpdf/Constants.h>
           ^~~~~~~~~~~~~~~~~~
  1 error generated.
  error: command 'cc' failed with exit status 1

  ----------------------------------------
  Failed building wheel for pikepdf
  Running setup.py clean for pikepdf
Failed to build pikepdf
Could not build wheels for pikepdf which use PEP 517 and cannot be installed directly

```
**Expected behavior**
Successful installation of ocrmypdf

**System:**

- OS: 11.1-RELEASE-p6 (in a jail)
- pkp: 1.10.5
- Python: 3.7.3
- PIP: pip 19.0.3 from /usr/local/lib/python3.7/site-packages/pip (python 3.7)

**Additional context**
The same issue occurs when trying to install ocrmypdf v8.2.2 v8.2.0 v8.1.0 v8.0.1 v8.0.0 v7.4.0.
The last version that installs successfully is v6.2.5

**Comment by jbarlow83** [comment]
The issue is getting qpdf and pikepdf to compile and install on FreeBSD
https://github.com/pikepdf/pikepdf/issues/6

For a long time the issue was that qpdf maintenance in FreeBSD had fallen far behind current versions and the maintainer was inactive. qpdf 8.4.0 is now available so pikepdf should be compilable.

In addition the ABI issue discussed at length in that thread might need this patch if the issue is actually `clang`:

```diff
iff --git a/setup.py b/setup.py
index 4dd02865..74167ec8 100644
--- a/setup.py
+++ b/setup.py
@@ -90,6 +90,7 @@ class BuildExt(build_ext):
             if (
                 has_flag(self.compiler, '-fvisibility=hidden')
                 and sys.platform != 'darwin'
+                and 'bsd' not in sys.platform
             ):
                 opts.append('-fvisibility=hidden')
         elif ct == 'msvc':
```

**Comment by fritz-hh** [comment]
Thank you a lot for your answer.
I will check if the patch helps.

Note; Upgrading to the latest version of ocrmypdf is currently not critical. I reached my main goal that was to upgrade to tesseract 4 (and ltsm) (that is available since end of march as a freebsd port). As ocrmypdf v6.2.5 is compatible with tesseract 4, i am currently more than happy.

**Comment by jbarlow83** [comment]
I succeeded in getting pikepdf to install on FreeBSD. 
https://pikepdf.readthedocs.io/en/latest/installation.html#installing-on-freebsd-11-2

The patch I proposed above is irrelevant. The bottom line is that `pip` on FreeBSD does not seem to check `/usr/local/include` for headers it needs for binary buildings, and FreeBSD also installs some system packages to that location. I added a kludge to setup.py to look in `/usr/local/include` for FreeBSD and that seems to do it.
